### PR TITLE
Use 'auto', range-'for', and structured bindings to improve some loops

### PR DESCRIPTION
**Pull request by vittorioromeo**
## Description

Refactoring change. Should not affect behavior at all. The `auto`, range-`for`, and structued bindings features have been used to improve the readability of some loops and usages of iterators.

## Tasks

* [ ] Tested on Linux
* [x] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

CI.

### CODE REVIEW

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: build this please.

File: src/SFML/Graphics/RenderTextureImplFBO.cpp
Comment (Bromeon) [inline src/SFML/Graphics/RenderTextureImplFBO.cpp:167]: Maybe use more expressive variable names, such as `contextId` and `frameBufferId` or so?
That's how they are named [here](https://github.com/SFML/SFML/blob/master/src/SFML/Graphics/RenderTextureImplFBO.cpp#L91-L93), for example.

File: examples/joystick/Joystick.cpp
Comment (Bromeon) [inline examples/joystick/Joystick.cpp:145]: Maybe change the names to `[label, joystick]` ("k/v" is not really more expressive than "first/second").

Also below.

**Review by Bromeon — COMMENTED** [review]
Thank you! I like how you kept the type name in the range-based for loop instead of `auto`-ing everything, probably helps the understanding.

File: src/SFML/Graphics/RenderTextureImplFBO.cpp
Comment (vittorioromeo) [reply src/SFML/Graphics/RenderTextureImplFBO.cpp:598]: Yeah, will change. Consistency is good :)

File: src/SFML/Graphics/RenderTextureImplFBO.cpp
Comment (vittorioromeo) [reply src/SFML/Graphics/RenderTextureImplFBO.cpp:167]: Good point, will change.

File: examples/joystick/Joystick.cpp
Comment (vittorioromeo) [reply examples/joystick/Joystick.cpp:145]: Ditto, will change.

File: src/SFML/Network/Http.cpp
Comment (vittorioromeo) [reply src/SFML/Network/Http.cpp:158]: I like the if-with-initializer, I would like to encourage its use. It might not be that useful here, but it's nice to start to be consistent with the feature. I will remove the `else` branch (redundant) as well, which does limit the scope of `it` in a useful way.

**Comment by vittorioromeo** [comment]
@Bromeon: I addressed your comments, please take a look :)

**Comment by ChrisThrasher** [comment]
I found a few more places to use ranged-for:

src/SFML/Window/Unix/WindowImplX11.cpp line 369
src/SFML/Window/Win32/JoystickImpl.cpp lines 468 and 1052

**Comment by vittorioromeo** [comment]
> I found a few more places to use ranged-for:
> 
> src/SFML/Window/Unix/WindowImplX11.cpp line 369 src/SFML/Window/Win32/JoystickImpl.cpp lines 468 and 1052

Thanks! I will handle those in a separate PR (or feel free to do that yourself), I'd like this one to be merged since CI is happy with it :)

**Comment by ChrisThrasher** [comment]
Sorry, I was looking at the master branch, not your branch when I referenced those files! I did find one loop you missed though at src/SFML/Network/Packet.cpp:578. I might submit a PR after yours get merged. Looks like a good first contribution for me.

File: src/SFML/Graphics/Image.cpp
Comment (Bromeon) [inline src/SFML/Graphics/Image.cpp:331]: I know it's technically correct, but this will always be `std::ptrdiff_t`, no?
Should we use that to help readability?

File: src/SFML/Graphics/RenderTextureImplFBO.cpp
Comment (Bromeon) [inline src/SFML/Graphics/RenderTextureImplFBO.cpp:63]: missing space before (

File: src/SFML/Graphics/RenderTextureImplFBO.cpp
Comment (Bromeon) [inline src/SFML/Graphics/RenderTextureImplFBO.cpp:598]: Maybe name the two `frameBufferIt` and `multisampleIt`?

File: src/SFML/Network/Http.cpp
Comment (Bromeon) [reply src/SFML/Network/Http.cpp:158]: I'm not really convinced that:
```cpp
if (auto it = m_fields.find(toLower(field)); it != m_fields.end())
{
     return it->second;
}

static const std::string empty = "";
return empty;
```
is more readable than:
```cpp
auto it = m_fields.find(toLower(field));
if (it != m_fields.end())
{
      return it->second;
}
else
{
     static const std::string empty = "";
     return empty;
}
```
or better, without negation and useless static:
```cpp
auto it = m_fields.find(toLower(field));
if (it == m_fields.end())
     return std::string();
else
     return it->second;
```

In the latter example, it's immediately clear that:
* the condition is to check an iterator against `end()`
* there are two outcomes (if/else) on which it is dispatched

Any other view on this, e.g. @eXpl0it3r? Could of course be more compact with `{}` and `?:`, but this is not necessarily what we're optimizing for 😉 

--- 

Apart from that, we should just return a default-constructed `std::string` instance, a reasonable implementation should not allocate memory.

**Review by Bromeon — COMMENTED** [review]
Thank you! Some minor comments:

File: src/SFML/Network/Http.cpp
Comment (eXpl0it3r) [reply src/SFML/Network/Http.cpp:158]: I guess readability for a new language/syntax feature is a bit tricky, as the good-old style of course will be "more readable" as we've been used to it for the past decade or longer.
I like the functionality that if-with-initializer provides, so I would at least try it out for a while before judging its readability.

I'm very much used to early-returns and dropping useless conditional branching (i.e. no `else` block after `return` in an `if` block). So that part would be fine for me.

Is the `static const std::string` really "useless"? As it would ensure that it's only created once, or am I wrong?

One alternative not yet mentioned would be:
```
auto  it = m_fields.find(toLower(field));
return it == m_fields.end() ? std::string() : it->second;
```

I have no strong opinions on any of the styles and don't know how much time we should invest in finding the most pleasing solution. 😄

Comment (vittorioromeo) [reply src/SFML/Network/Http.cpp:158]: I don't think we should focus on the style of this code much. All mentioned alternatives are quite readable. I prefer my own because it limits the scope of the iterator to the `if`, and reducing scope of variables is IMHO a good thing.

Also, we are returning `const std::string&', so we cannot return a temporary.

File: src/SFML/Graphics/Image.cpp
Comment (vittorioromeo) [reply src/SFML/Graphics/Image.cpp:331]: Hmm, I wouldn't risk it to be honest. We support many platforms and I would have to ensure that all of them do indeed use `std::ptrdiff_t`. I'd rather be "technically correct" in this case.

File: src/SFML/Graphics/RenderTextureImplFBO.cpp
Comment (vittorioromeo) [reply src/SFML/Graphics/RenderTextureImplFBO.cpp:63]: Will fix

File: src/SFML/Graphics/RenderTextureImplFBO.cpp
Comment (vittorioromeo) [reply src/SFML/Graphics/RenderTextureImplFBO.cpp:598]: Will fix

**Comment by vittorioromeo** [comment]
@Bromeon: I think this can be merged if you are happy with the latest changes.

File: src/SFML/Network/Http.cpp
Comment (Bromeon) [reply src/SFML/Network/Http.cpp:158]: Regarding if-with-initializer: it can make sense if we have a bigger scope and we want to keep lifetime of variables as local as possible (this may also be an indicator for a separate method). However, none of the advantages play out here:
* it's not a big scope
* there is no code coming after, which could accidentally use the variable
* it's not an early return, it's the last statement in the function

I agree it's not worth discussing for hours, but maybe precisely for that reason we should address the larger issues before getting into bikeshedding territory 🙂  we can leave this particular change in, but please let's not hunt down `if` statements in SFML now 😆 

---

> Also, we are returning const std::string&, so we cannot return a temporary. 
> We are also not allocating any memory by creating an empty static std::string.

True, I missed that we return a const-reference.

We're not dynamically allocating memory, but we reserve static memory until the end of the program, even if this method is only used once. Plus, we add potential thread unsafety. Both are very minor issues here, but keep in mind that they're trade-offs for another minor issue (copying a `std::string` in a code that likely suffers from much larger performance impact, network latency). 

Anyway, this just as an aside 🙂

File: src/SFML/Window/OSX/JoystickImpl.cpp
Comment (Bromeon) [inline src/SFML/Window/OSX/JoystickImpl.cpp:365]: There's no way to ignore a destructuring binding (here `iohidElementRef`), but doesn't this cause unused-variable warnings?

Unfortunately, `std::ignore` only works with `std::tie()`, which cannot be used for initialization. ([some SO discussion](https://stackoverflow.com/q/40673080))

**Review by Bromeon — COMMENTED** [review]
Thanks again!

Looks good, just 1 comment about potential warning. If that is not an issue, we can go ahead 🙂

File: src/SFML/Window/OSX/JoystickImpl.cpp
Comment (vittorioromeo) [reply src/SFML/Window/OSX/JoystickImpl.cpp:365]: There's really no way to ignore a single binding in a destructed binding declaration. This is a (known) annoying limitation.

Comment (Bromeon) [reply src/SFML/Window/OSX/JoystickImpl.cpp:365]: There was a [SO comment](https://stackoverflow.com/questions/40673080/stdignore-with-structured-bindings#comment76711644_40680014) saying it would generate a warning, but maybe that's not accurate or no longer the case.

I'm happy if this passes CI, no need to create problems where there are none 🙂

Comment (vittorioromeo) [reply src/SFML/Window/OSX/JoystickImpl.cpp:365]: CI seems happy, let's merge!

Comment (vittorioromeo) [reply src/SFML/Window/OSX/JoystickImpl.cpp:365]: @Bromeon ping :)

**Review by Bromeon — APPROVED** [review]

**Pull request merged**
